http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala

```diff
@@ -528,9 +540,127 @@ class LeoPubsubMessageSubscriber[F[_]: Timer: ContextShift](
       )
     } yield ()
 
-  def handleCreateAppMessage(msg: CreateAppMessage): F[Unit] = F.unit
+  def handleCreateAppMessage(msg: CreateAppMessage)(
+    implicit ev: ApplicativeAsk[F, AppContext]
+  ): F[Unit] =
+    for {
+      ctx <- ev.ask
+      initialGoogleCall = msg.cluster.traverse(createClusterMessage =>
+        gkeInterp.createCluster(createClusterMessage, msg.appId)
+      )
+      clusterResult <- initialGoogleCall.onError {
+        case e =>
+          for {
+            ctx <- ev.ask
+            _ <- logger.error(e)(s"Failed to delete app for app ${msg.appId}")
+            _ <- appErrorQuery
+              .save(msg.appId,
+                    KubernetesError(e.getMessage(), ctx.now, ErrorAction.CreateGalaxyApp, ErrorSource.Cluster, None))
+              .transaction
+          } yield ()
+      }
+      createClusterAndNodepool = for {
+        _ <- clusterResult.fold(F.unit)(result => gkeInterp.pollCluster(result))
+        _ <- gkeInterp.createAndPollNodepool(msg.nodepoolId, msg.appId)
+      } yield ()
+      diskCreation = createDiskForApp(msg)
+      parPreAppCreationSetup = List(createClusterAndNodepool, diskCreation).parSequence.flatMap(_ => F.unit)
+
+      task = for {
+        _ <- parPreAppCreationSetup
+        _ <- gkeInterp.createAndPollApp(msg)
+      } yield ()
 
-  def handleDeleteAppMessage(msg: DeleteAppMessage): F[Unit] = F.unit
+      _ <- asyncTasks.enqueue1(
+        Task(ctx.traceId, task, Some(logError(s"${ctx.traceId.asString} | ${msg.appId}", "creating app")), ctx.now)
+      )
+    } yield ()
+
+  private def deleteDiskForApp(diskId: DiskId, appId: AppId)(
+    implicit ev: ApplicativeAsk[F, AppContext]
+  ): F[Unit] = {
+    val delete = for {
+      ctx <- ev.ask
+      _ <- handleDeleteDiskMessage(DeleteDiskMessage(diskId, Some(ctx.traceId)), true)
+    } yield ()
+
+    delete.onError {
+      case e =>
+        for {
+          ctx <- ev.ask
+          _ <- logger.error(e)(s"Failed to delete disk for app ${appId}")
+          _ <- appErrorQuery
```

Comment:
what exactly is the difference? The reason that any failures are surrounded by onError is because we care what step of the process the error occurred, as we want to save the `ErrorSource`

Reply:
now that I think about it, I think persisting to DB should happen in https://github.com/DataBiosphere/leonardo/blob/develop/http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala#L83...

To your question, it removes some duplicate code, and we won't have db save and log.error everywhere randomly. You would still know the source because you'll do `adaptError(KuberErro(xxx, errorSource = ???, ???))`